torch_xla/experimental/fori_loop.py

```diff
@@ -12,40 +12,70 @@
 from torch._higher_order_ops.while_loop import while_loop_op
 
 
+def fori_loop(lower, upper, body_fun, one_value, init_val):
+
+  device = xm.xla_device()
+
+  def cond_fn(upper, lower, x):
+    return lower[0] < upper[0]
+
+  def body_fn(upper, lower, x):
+    one_value = torch.ones(1, dtype=torch.int32, device=device)
+    return (torch.sub(upper, one_value), lower, body_fun(one_value, x))
+
+  res = while_loop(cond_fn, body_fn, (lower, upper, init_val))
+  return res
+
+
 @while_loop_op.py_impl(DispatchKey.XLA)
-def while_loop(cond_fn, body_fn, carried_inputs, additional_inputs=None):
+def while_loop(cond_fn, body_fn, *carried_inputs, additional_inputs=None):
+  # TODO(@manfei): PyTorch require carried_inputs to be list/tuple, PyTorch/XLA _xla_while_loop only accept *operands, *operands would tuple items again: (a, '')
   # cond_fn&body_fn: callable
   # carried_inputs: (Tuple of possibly nested dict/list/tuple of tensors)
   if additional_inputs is None:
     additional_inputs = tuple()
-  return _xla_while_loop(cond_fn, body_fn, carried_inputs, additional_inputs)
+  return _xla_while_loop(
+      cond_fn, body_fn, *carried_inputs, additional_inputs=additional_inputs)
 
 
-def _xla_while_loop(cond_fn, body_fn, carried_inputs, additional_inputs):
+def _xla_while_loop(cond_fn, body_fn, *carried_inputs, additional_inputs):
+  # untuple carried_inputs from while_loop
+  carried_inputs = carried_inputs[0]
+  # fake carried_inputs to split formal code
```

Comment:
@JackCaoG could you help me understand this part? why we need this.

Reply:
more context: this part create fake tensor from input tensors, so that fake tensor would not include formal IR like input tensors, and then we use clean fake tensor to create body_xlacomputation/cond_xlacomputation